vllm/utils.py

```diff
@@ -447,13 +447,25 @@ def is_pin_memory_available() -> bool:
 
 class CudaMemoryProfiler:
 
-    def __init__(self, device=None):
+    def __init__(self, device=None, capture_max_memory: bool = False):
+        """A context manager to measure memory usage on a given device.
+
+        If capture_max_memory is True, it measures the maximum memory usage
+        during the profiling. However, it can only measure GPU memory used by
+        torch tensor. If it is False, it measures the memory delta which also
+        includes non-torch tensor GPU memory usage.
+        """
         self.device = device
+        self.capture_max_memory = capture_max_memory
 
     def current_memory_usage(self) -> float:
         # Return the memory usage in bytes.
-        torch.cuda.reset_peak_memory_stats(self.device)
-        mem = torch.cuda.max_memory_allocated(self.device)
+        if self.capture_max_memory:
+            torch.cuda.reset_peak_memory_stats(self.device)
```

Comment:
IIUC, the change made in this PR is incorrect because `torch.cuda.mem_get_info` does not consider the "free" GPU memory managed by PyTorch caching allocator.

Once the GPU memory is allocated (via `cuda-malloc`) by PyTorch caching allocator, the memory is never `cuda-free`d unless the user enforces it (e.g., by `empty_cache`). While this memory can be regarded free by PyTorch allocator because it's not used for any tensor, it's not regarded free from the external point of view (e.g., in nvidia-smi or `torch.cuda.mem_get_info`). Therefore, the profiler cannot capture the GPU memory usage inside the PyTorch allocator, and thus may under-estimate the memory usage.

Reply:
ah interesting... so it is like buffered memory. 

one issue was that using the original method couldn't take into account of memory used by non-torch. Let me see if I can find a way to take into account of buffer.

Reply:
Currently thinking about 2 different options;

- **Option 1**: Keep the current memory context manager as it is. For cuda graph, we measure the memory using snapshot (using mem_get_info). for kv cache and model, it should be fine measuring the memory using the status quo (a.k.a., memory_allocated because gpu memory is used just for tensors in these cases)
- **Option 2**: Modify the memory context manager to use
```
free_mem_from_torch = torch.cuda.memory_reserved (total mem used by allocator) - torch.cuda.memory_allocated (total mem used by tensor). 
free, total = torch.cuda.mem_get_info(self.device)
mem_usage = total - (free + free_mem_from_torch)
```

This should include releasable memory from torch cache allocator to the free memory (based on https://pytorch.org/docs/stable/notes/cuda.html#memory-management)

Reply:
> For cuda graph, we measure the memory using snapshot (using mem_get_info).

What do you exactly mean by this?

Reply:
So keep other part as it is. And for cuda graph capture part, we measure memoroy by 
```python3
free, total = torch.cuda.mem_get_info(self.device)
used = total - free
cuda_capture()
free, total = torch.cuda.mem_get_info(self.device)
used_after = total - free
memory_used = used_after - used
```